### PR TITLE
[FLINK-36529] Allow Flink version configs to be set to greater than given version

### PR DESCRIPTION
**Pull request by tomncooper**
## What is the purpose of the change

The operator currently allows the following syntax for defining flink version specific defaults:

`kubernetes.operator.default-configuration.flink-version.v1_18.key: value`

The problem with this is that, in many cases, these defaults should be applied to newer Flink versions as well, forcing config duplications.

This PR introduces a new "greater than" syntax for config defaults, indicating that they should be applied to a given version and above:
`kubernetes.operator.default-configuration.flink-version.v1_18+.key: value`

In this case key:value would be applied to all Flink version greater or equal to 1.18, unless overridden for specific versions.

## Brief change log

- Adds a new method `getRelevantVersionPrefixes`, to the `org.apache.flink.kubernetes.operator.config.FlinkConfigManager`, which identifies all Flink version default config prefixes which are relevant to the currently specified Flink version. These are then saved in cache (map), keyed on Flink version.
- Refactored `getDefaultConfig` to call the cached flink version prefixes or generated them if none exist.
- Clear the prefix cache when `updateDefaultConfig` is called. This ensures new prefixes will be calculated on the next call to `getDefaultConfig`. 
- Refactored the `FlinkVersion` enum to specify major and minor semver integers to facilitate quick look up of relevant Flink versions when parsing version strings.

## Verifying this change

This change added tests and can be verified as follows:

- Added additional tests in the `FlinkConfigManagerTest` class to cover the new Regex and `getRelevantVersionPrefixes` methods.
- Updated the `testVersionNamespaceDefaultConfs` test in `FlinkConfigManagerTest` to test the greater than version behaviour.
- Added a `FlinkVersionTest` to ensure refactoring of the enum was correct.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changes to the `CustomResourceDescriptors`: no
  - Core observer or reconciler logic that is regularly executed: yes

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? Update the configuration docs

### CODE REVIEW

**Comment by tomncooper** [comment]
I probably need to look at adding an end to end test for this but want to make sure I am on the right track before doing that.

**Comment by tomncooper** [comment]
CC @gyfora

**Comment by gyfora** [comment]
I will try to review this in detail tomorrow but scrolling through it, it look pretty good already

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java [hunk 1]
```diff
@@ -202,6 +215,65 @@ public FlinkOperatorConfiguration getOperatorConfiguration(
...
+                                                .FLINK_VERSION_GREATER_THAN_SUFFIX
+                                        + ".");
+                    }
+                } catch (NumberFormatException numberFormatException) {
+                    LOG.warn("Unable to parse version number in config key: {}", entry.getKey());
+                } catch (IllegalArgumentException illegalArgumentException) {
+                    LOG.warn("Unknown Flink version in config key: {}", entry.getKey());
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (tomncooper) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java:262]: Good point, we could generate the lists in the `FlinkVersion` enum.

[the remark this answers is not in the document]
Comment (tomncooper) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java:262]: Do you mean, for every supported version create all possible prefixes (for example we could do this in the `FlinkVersion` enum) and then loop through them for the given `FlinkVersion`? That would simplify things, but we would have wasted calls to `applyDefault` which scans the whole base config each time.

[the remark this answers is not in the document]
Comment (tomncooper) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java:262]: The current implementation scans `baseConfig` once and only creates prefixes for version that are there. Of course I could refactor `applyDefault` to accept a list of prefixes and check each config against them. Which is may be less checks overall.

[the remark this answers is not in the document]
Comment (tomncooper) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkConfigManager.java:269]: So I added a cache (map) in the config manager (keyed on flink version) and call that cache inside `getDefaultConfig`. I clear the cache in `updateDefaultConfig` so we only do the calc once for each FlinkVersion:baseConfig combination.

**Comment by tomncooper** [comment]
@gyfora Looks like the CI was cancelled rather than failed?

**Comment by gyfora** [comment]
There is an error in the CI
```
Please generate the java doc via 'mvn clean install -DskipTests -Pgenerate-docs' again
```

**Pull request merged**